expression/constant_propagation.go

```diff
@@ -241,20 +247,26 @@ func (s *propConstSolver) pickNewEQConds(visited []bool) (retMapper map[int]*Con
 		if visited[i] {
 			continue
 		}
-		col, con := validEqualCond(cond)
+		col, con := validEqualCond(s.ctx, cond)
 		// Then we check if this CNF item is a false constant. If so, we will set the whole condition to false.
 		var ok bool
 		if col == nil {
-			if con, ok = cond.(*Constant); ok {
-				value, _, err := EvalBool(s.ctx, []Expression{con}, chunk.Row{})
-				if err != nil {
-					terror.Log(err)
-					return nil
-				}
-				if !value {
-					s.setConds2ConstFalse()
-					return nil
-				}
+			con, ok = cond.(*Constant)
+			if !ok {
+				continue
+			}
+			visited[i] = true
+			if s.ctx.GetSessionVars().StmtCtx.UseCache && (con.DeferredExpr != nil || con.ParamMarker != nil) {
```

Comment:
This predicate is presented several times in this PR. It's better to extract a function for it to follow the [DRY(Don't Repeat Yourself)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle.